relayer/checkpoint.go

```diff
@@ -0,0 +1,212 @@
+}
+
+func (cm *checkpointManager) writeToDatabase() {
+	cm.lock.RLock()
+	defer cm.lock.RUnlock()
+	// Ensure we're not writing the default value
+	if cm.committedHeight == 0 {
+		return
+	}
+	storedHeight, err := getLatestProcessedBlockHeight(cm.database, cm.relayerID)
+	if err != nil && !database.IsKeyNotFoundError(err) {
+		cm.logger.Error(
+			"Failed to get latest processed block height",
+			zap.Error(err),
+			zap.String("relayerID", cm.relayerID.ID.String()),
+		)
+		return
+	}
+	if storedHeight >= cm.committedHeight {
+		return
+	}
+	cm.logger.Debug(
+		"Writing height",
+		zap.Uint64("height", cm.committedHeight),
+		zap.String("relayerID", cm.relayerID.ID.String()),
+	)
+	err = cm.database.Put(cm.relayerID.ID, database.LatestProcessedBlockKey, []byte(strconv.FormatUint(cm.committedHeight, 10)))
+	if err != nil {
+		cm.logger.Error(
+			"Failed to write latest processed block height",
+			zap.Error(err),
+			zap.String("relayerID", cm.relayerID.ID.String()),
+		)
+		return
+	}
+}
+
+func (cm *checkpointManager) listenForWriteSignal() {
+	for range cm.writeSignal {
+		cm.writeToDatabase()
+	}
+}
+
+func (cm *checkpointManager) incrementFinishedCounter(height uint64) {
+	cm.lock.Lock()
+	defer cm.lock.Unlock()
+	counter, ok := cm.queuedHeightsAndMessages[height]
+	if !ok {
+		// This is expected for manual Warp messages, since they are not associated with a height
+		// and are therefore not tracked by the checkpoint manager.
+		cm.logger.Debug(
+			"Pending height not found",
+			zap.Uint64("height", height),
+			zap.String("relayerID", cm.relayerID.ID.String()),
+		)
+		return
+	}
+
+	counter.processedMessages++
+	cm.logger.Debug(
+		"Received finished signal",
+		zap.Uint64("height", height),
+		zap.String("relayerID", cm.relayerID.ID.String()),
+		zap.Uint64("processedMessages", counter.processedMessages),
+		zap.Uint64("totalMessages", counter.totalMessages),
+	)
+	if counter.processedMessages == counter.totalMessages {
+		cm.stageCommittedHeight(height)
+		delete(cm.queuedHeightsAndMessages, height)
+	}
+}
+
+// handleFinishedRelays listens for finished signals from the application relayer, and commits the
+// height once all messages have been processed.
+// This function should only be called once.
+func (cm *checkpointManager) listenForFinishedRelays() {
+	for height := range cm.finished {
+		cm.incrementFinishedCounter(height)
+	}
+}
+
+// stageCommittedHeight queues a height to be written to the database.
+// Heights are committed in sequence, so if height is not exactly one
+// greater than the current committedHeight, it is instead cached in memory
+// to potentially be committed later.
+// Requires that cm.lock be held
+func (cm *checkpointManager) stageCommittedHeight(height uint64) {
+	if cm.committedHeight == 0 {
+		cm.logger.Debug(
+			"Committing initial height",
+			zap.Uint64("height", height),
+			zap.String("relayerID", cm.relayerID.ID.String()),
+		)
+		cm.committedHeight = height
+		return
+	}
+
```

Comment:
We actually need to ensure that `height` > `cm.committedHeight` here. I think with the race condition i described above it's currently possible. I think it should at least emit an error log, as it means we've probably made a logic error here. A `height` <= `cm.committedHeight` would mean we would never pop anything from the heap.

Reply:
See https://github.com/ava-labs/awm-relayer/pull/263#discussion_r1585291255